src/hotspot/share/runtime/handshake.cpp

```diff
@@ -587,3 +596,133 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
                        pr_ret == HandshakeState::_succeeded ? "including" : "excluding", p2i(match_op));
   return pr_ret;
 }
+
+void HandshakeState::lock() {
+  _lock.lock_without_safepoint_check();
+}
+
+void HandshakeState::unlock() {
+  _lock.unlock();
+}
+
+// The careful dance between thread suspension and exit is handled here.
+// The caller has notified the agents that we are exiting, before we go
+// on, we must check for a pending external suspend request and honor it
+// in order to not surprise the thread that made the suspend request.
+void HandshakeState::thread_exit() {
+  guarantee(JavaThread::current() == _handshakee, "Must be");
+  while (true) {
+    {
+      MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+      if (!is_suspend_requested()) {
+        _handshakee->set_exiting();
+        // No more external suspends are allowed at this point.
+        return;
+      }
+    }
+    ThreadBlockInVM tbivm(_handshakee);
+    // We're done with this suspend request, but we have to loop around
+    // and check again. Eventually we will get the lock without a pending
+    // external suspend request and will be able to mark ourselves as
+    // exiting.
+  }
+}
+
+bool HandshakeState::suspend_request_pending() {
+  assert(JavaThread::current()->thread_state() != _thread_blocked, "should not be in a blocked state");
+  assert(JavaThread::current()->thread_state() != _thread_in_native, "should not be in native");
```

Comment:
I missed one assert, this must be asked on yourself.
If you ask while in blocked/native the answer may change at any moment, since processing of handshake is ok.
To get a stable result you must be in another state.

Note @pchilano think we should remove this method, so either I add assert and comment or remove it.